File: sdks/local-rp/java/src/main/java/community/catalyst/linkkeys/localrp/SdkException.java
package community.catalyst.linkkeys.localrp;

/**
 * The SDK's network/IO-layer error type &mdash; everything that isn't a pure
 * protocol verification failure ({@link LocalRpError}/{@link ClaimError}/
 * {@link RevocationError}). Every fallible network operation in this SDK
 * throws one of these. None of these variants carry key material, nonces,
 * tokens, tickets, or claim values (AGENTS.md: "Never log sensitive
 * information") &mdash; only domain names, field names, and short messages.
 */
public class SdkException extends RuntimeException {
    public enum Kind {
        /** A field the caller supplied was structurally invalid. */
        INVALID_INPUT,
        /** DNS TXT lookup or record parsing failed for a domain. */
        DNS,
        /** The TCP transport could not reach a domain's endpoint. */
        TRANSPORT,
        /** TLS handshake / certificate pinning failed. */
        TLS,
        /** The CSIL-RPC envelope could not be encoded/decoded, or wire framing was malformed. */
        PROTOCOL,
        /** The peer returned a non-Ok RPC transport status. */
        SERVER,
        /** No trustworthy domain keys were established for a domain. */
        NO_TRUSTED_DOMAIN_KEYS,
        /**
         * Fetching or decoding a domain's revocation list failed. Fatal
         * (fail closed): revocation delivery is mandatory context for
         * trusting a domain's keys, not a best-effort nicety &mdash; a
         * hostile or merely-unreachable IDP must not be able to suppress a
         * key's revocation by dropping or erroring this call.
         */
        REVOCATION_UNAVAILABLE,
    }

    private final Kind kind;
    private final int serverStatus;

    public SdkException(Kind kind, String message) {
        super(message);
        this.kind = kind;
        this.serverStatus = 0;
    }

    public SdkException(Kind kind, String message, Throwable cause) {
        super(message, cause);
        this.kind = kind;
        this.serverStatus = 0;
    }

    public SdkException(int serverStatus, String message) {
        super("server error (" + serverStatus + "): " + message);
        this.kind = Kind.SERVER;
        this.serverStatus = serverStatus;
    }

    public Kind kind() {
        return kind;
    }

    public int serverStatus() {
        return serverStatus;
    }
}

File: sdks/local-rp/csharp/src/LinkKeys.LocalRp/SdkException.cs
namespace LinkKeys.LocalRp;

/// <summary>
/// The SDK's network/IO-layer error type — everything that isn't a pure protocol
/// verification failure (<see cref="LocalRpError"/>/<see cref="ClaimError"/>/
/// <see cref="RevocationError"/>). Every fallible network operation in this SDK throws
/// one of these. None of these variants carry key material, nonces, tokens, tickets, or
/// claim values (AGENTS.md: "Never log sensitive information") — only domain names,
/// field names, and short messages.
/// </summary>
public class SdkException : Exception
{
    public enum ErrorKind
    {
        /// <summary>A field the caller supplied was structurally invalid.</summary>
        InvalidInput,

        /// <summary>DNS TXT lookup or record parsing failed for a domain.</summary>
        Dns,

        /// <summary>The TCP transport could not reach a domain's endpoint.</summary>
        Transport,

        /// <summary>TLS handshake / certificate pinning failed.</summary>
        Tls,

        /// <summary>The CSIL-RPC envelope could not be encoded/decoded, or wire framing was malformed.</summary>
        Protocol,

        /// <summary>The peer returned a non-Ok RPC transport status.</summary>
        Server,

        /// <summary>No trustworthy domain keys were established for a domain.</summary>
        NoTrustedDomainKeys,

        /// <summary>
        /// <c>DomainKeys/get-revocations</c> could not be fetched or decoded for a
        /// domain. Fatal (fail closed): revocation delivery is mandatory context for
        /// trusting a domain's keys at all, so an unreachable/erroring/malformed
        /// get-revocations call must never be silently treated as "no revocations" — a
        /// hostile or merely-unreliable IDP could otherwise suppress a key's revocation
        /// simply by dropping this call.
        /// </summary>
        RevocationUnavailable,
    }

    public ErrorKind Kind { get; }

    public int ServerStatus { get; }

    public SdkException(ErrorKind kind, string message) : base(message)
    {
        Kind = kind;
        ServerStatus = 0;
    }

    public SdkException(ErrorKind kind, string message, Exception cause) : base(message, cause)
    {
        Kind = kind;
        ServerStatus = 0;
    }

    public SdkException(int serverStatus, string message) : base($"server error ({serverStatus}): {message}")
    {
        Kind = ErrorKind.Server;
        ServerStatus = serverStatus;
    }
}
